### PR TITLE
Generalize service implementations

### PR DESCRIPTION
**Pull request by darinkrauss**
- Implement general ServicesManager and service serialization/restoration
- Replace OSLog and DiagnosticLogger with DiagnosticLog and connect to LoggingManager
- Move GlucoseThreshold, LoopSettings, and TempBasalRecommendation to LoopKit
- Move NightscoutUploader to NightscoutService
- Move Amplitude, Loggly, and Nightscout service implementations to respective repos
- Implement generalized LoggingManager, AnalyticsManager, and RemoteDataManager to connect to specific services
- Refactor Analytics singleton to injectable implementation
- Remove unused MLabService KeyChain extension
- Refactor SettingsTableViewController to generalize service setup

https://tidepool.atlassian.net/browse/LOOP-11
https://tidepool.atlassian.net/browse/LOOP-13
https://tidepool.atlassian.net/browse/LOOP-10
https://tidepool.atlassian.net/browse/LOOP-27
https://tidepool.atlassian.net/browse/LOOP-22

### CODE REVIEW

File: Loop/AppDelegate.swift [hunk 1]
```diff
@@ -8,28 +8,41 @@
...
 final class AppDelegate: UIResponder, UIApplicationDelegate {
 
-    private lazy var log = DiagnosticLogger.shared.forCategory("AppDelegate")
+    private lazy var log = DiagnosticLog(category: "AppDelegate")
+
+    private lazy var servicesManager = ServicesManager()
+
+    private lazy var analyticsManager = AnalyticsManager(servicesManager: servicesManager)
+
+    private lazy var loggingManager = LoggingManager(servicesManager: servicesManager)
```
Comment (nhamming) [inline Loop/AppDelegate.swift]: Should some of these be constants?

File: Loop/Managers/RemoteDataManager.swift [hunk 2]
```diff
@@ -8,34 +8,177 @@
...
-    private let keychain = KeychainManager()
+    func upload(glucoseValues values: [GlucoseValue], sensorState: SensorDisplayable?) {
+        remoteData.forEach { $0.upload(glucoseValues: values, sensorState: sensorState) }
+    }
+
+    func upload(pumpEvents events: [PersistedPumpEvent], fromSource source: String, completion: @escaping (Result<[URL], Error>) -> Void) {
+        // TODO: How to handle completion correctly
+        if remoteData.count > 0 {
+            remoteData[0].upload(pumpEvents: events, fromSource: source, completion: completion)
+        }
```
Comment (nhamming) [inline Loop/Managers/RemoteDataManager.swift]: Why only the first element? Similar comment for code below.

File: Loop/Managers/Service.swift [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import LogglyServiceKit
+import NightscoutServiceKit
+
+
+/// The order here specifies the order in the service selection popup
+let serviceTypes: [Service.Type] = [
+    NightscoutService.self,
+    LogglyService.self,
+    AmplitudeService.self,
+    MockService.self,
```
Comment (nhamming) [inline Loop/Managers/Service.swift:21]: Should the MockService be limited to just debug builds?

**Review by nhamming — APPROVED** [review]
That is all.

File: Loop/Managers/RemoteDataManager.swift (hunk 2 above)
Comment (darinkrauss) [reply Loop/Managers/RemoteDataManager.swift]: This is a placeholder that will only work for one `RemoteData` service at a time. However, at this point with the code there is only one `RemoteData` service anyways, namely `NightscoutService`. 

This is one reason why we are moving from the current "push" model to a "pull" model. The "pull" model will allow multiple `RemoteData` services to be active at the same time (eg. `NightscoutService` and `TidepoolService`).

So, this is a "good enough for now" situation that will be updated with the next round of updates to the `RemoteData` service and `NightscoutService`.

File: Loop/Managers/Service.swift (hunk 3 above)
Comment (darinkrauss) [reply Loop/Managers/Service.swift:21]: Yes. Will make that change.

File: Loop/AppDelegate.swift (hunk 1 above)
Comment (darinkrauss) [reply Loop/AppDelegate.swift]: I believe a `lazy` property must be a `var` (not `let`).

**Review by nhamming — APPROVED** [review]
All good!

**Review by ps2 — APPROVED** [review]
🚀 Looks good!

**Pull request merged**